File: connectors/javascript/tests/integration/advanced/A05_MultipleConnector.test.ts
/**
 * A-05: Multiple Connector Tests
 *
 * Tests for using multiple Connector instances simultaneously.
 * Verifies independent connections and communication.
 * Useful for multi-server connections or simulating multiple clients in tests.
 */

import { describe, test, expect, beforeEach, afterEach } from 'vitest';
import { Connector } from '../../../src/connector.js';
import { Packet } from '../../../src/packet.js';
import { TestServerClient, CreateStageResponse } from '../helpers/TestServerClient.js';
import { serializeAuthenticateRequest, parseAuthenticateReply, serializeEchoRequest, parseEchoReply, serializeBroadcastRequest, parseBroadcastNotify } from '../helpers/TestMessages.js';

describe('A-05: Multiple Connectors', () => {
    const testServer = new TestServerClient();
    const connectors: Connector[] = [];
    const stages: CreateStageResponse[] = [];

    beforeEach(async () => {
        // Create 5 independent stages
        for (let i = 0; i < 5; i++) {
            stages.push(await testServer.createStage());
        }
    }, 30000);

    afterEach(async () => {
        // Cleanup all connectors
        for (const connector of connectors) {
            if (connector.isConnected) {
                connector.disconnect();
                await delay(50);
            }
        }
        connectors.length = 0;
        stages.length = 0;
    });

    function createConnector(requestTimeoutMs = 10000): Connector {
        const connector = new Connector();
        connector.init({
            requestTimeoutMs,
            heartbeatIntervalMs: 10000
        });
        connectors.push(connector);
        return connector;
    }

    async function delay(ms: number): Promise<void> {
        return new Promise(resolve => setTimeout(resolve, ms));
    }

    async function authenticate(conn: Connector, userId: string): Promise<any> {
        const payload = serializeAuthenticateRequest({ userId, token: `token-${userId}` });
        const authRequest = Packet.fromBytes('AuthenticateRequest', payload);
        const response = await conn.authenticate(authRequest);
        return parseAuthenticateReply(response.payload);
    }

    test('A-05-01: Multiple connectors can connect simultaneously', async () => {
        // Given: 5 connectors
        const testConnectors = Array.from({ length: 5 }, () => createConnector());

        // When: Connect all simultaneously
        const connectPromises = testConnectors.map((connector, index) =>
            connector.connect(testServer.wsUrl, stages[index].stageId, stages[index].stageType)
        );

        const results = await Promise.all(connectPromises);

        // Then: All connections should succeed
        expect(results.every(r => r === true)).toBe(true);
        expect(testConnectors.every(c => c.isConnected)).toBe(true);
    });

    test('A-05-02: Multiple connectors can authenticate independently', async () => {
        // Given: 3 connected connectors
        const testConnectors = Array.from({ length: 3 }, () => createConnector());

        for (let i = 0; i < testConnectors.length; i++) {
            await testConnectors[i].connect(
                testServer.wsUrl,
                stages[i].stageId,
                stages[i].stageType
            );
        }

        // When: Authenticate each with different user
        const authPromises = testConnectors.map(async (connector, index) => {
            return await authenticate(connector, `multi-user-${index}`);
        });

        const authResults = await Promise.all(authPromises);

        // Then: All authentications should succeed
        expect(testConnectors.every(c => c.isAuthenticated)).toBe(true);
        authResults.forEach((result, index) => {
            expect(result).toBeDefined();
            expect(result.success).toBe(true);
        });
    });

    test('A-05-03: Multiple connectors can send requests simultaneously', async () => {
        // Given: 3 connected and authenticated connectors
        const testConnectors = Array.from({ length: 3 }, () => createConnector());

        for (let i = 0; i < testConnectors.length; i++) {
            await testConnectors[i].connect(
                testServer.wsUrl,
                stages[i].stageId,
                stages[i].stageType
            );

            await authenticate(testConnectors[i], `multi-user-${i}`);
        }

        // When: Send echo requests simultaneously
        const requestPromises = testConnectors.map(async (connector, index) => {
            const payload = serializeEchoRequest({ content: `Echo ${index}`, sequence: index });
            const echoRequest = Packet.fromBytes('EchoRequest', payload);
            return await connector.request(echoRequest);
        });

        const responses = await Promise.all(requestPromises);

        // Then: All requests should complete
        expect(responses).toHaveLength(3);
        responses.forEach((response, index) => {
            expect(response.msgId).toBe('EchoReply');
            const reply = parseEchoReply(response.payload);
            expect(reply.content).toBe(`Echo ${index}`);
            expect(reply.sequence).toBe(index);
        });
    });

    test('A-05-04: Disconnecting one connector does not affect others', async () => {
        // Given: 3 connected and authenticated connectors
        const testConnectors = Array.from({ length: 3 }, () => createConnector());

        for (let i = 0; i < testConnectors.length; i++) {
            await testConnectors[i].connect(
                testServer.wsUrl,
                stages[i].stageId,
                stages[i].stageType
            );

            await authenticate(testConnectors[i], `multi-user-${i}`);
        }

        // When: Disconnect first connector
        testConnectors[0].disconnect();
        await delay(500);

        // Then: First is disconnected, others remain connected
        expect(testConnectors[0].isConnected).toBe(false);
        expect(testConnectors[1].isConnected).toBe(true);
        expect(testConnectors[2].isConnected).toBe(true);

        // Others can still make requests
        for (let i = 1; i < testConnectors.length; i++) {
            const payload = serializeEchoRequest({ content: `Echo ${i}`, sequence: i });
            const echoRequest = Packet.fromBytes('EchoRequest', payload);
            const response = await testConnectors[i].request(echoRequest);
            expect(response.msgId).toBe('EchoReply');
            const reply = parseEchoReply(response.payload);
            expect(reply.content).toBe(`Echo ${i}`);
        }
    });

    test('A-05-05: Multiple connectors can connect to same stage', async () => {
        // Given: Same stage for all connectors
        const sharedStage = stages[0];
        const testConnectors = Array.from({ length: 3 }, () => createConnector());

        // When: Connect all to same stage
        for (let i = 0; i < testConnectors.length; i++) {
            await testConnectors[i].connect(
                testServer.wsUrl,
                sharedStage.stageId,
                sharedStage.stageType
            );

            await authenticate(testConnectors[i], `multi-user-${i}`);
        }

        // Then: All can make independent requests
        for (let i = 0; i < testConnectors.length; i++) {
            const payload = serializeEchoRequest({ content: `Echo ${i}`, sequence: i });
            const echoRequest = Packet.fromBytes('EchoRequest', payload);
            const response = await testConnectors[i].request(echoRequest);
            expect(response.msgId).toBe('EchoReply');
            const reply = parseEchoReply(response.payload);
            expect(reply.content).toBe(`Echo ${i}`);
        }
    });

    test('A-05-06: Stress test with many connectors', async () => {
        // Given: 10 connectors
        const connectorCount = 10;
        const testConnectors: Connector[] = [];
        const extraStages: CreateStageResponse[] = [];

        // Create additional stages
        for (let i = 0; i < connectorCount; i++) {
            extraStages.push(await testServer.createStage());
        }

        // When: Connect all simultaneously
        const connectPromises: Promise<boolean>[] = [];
        for (let i = 0; i < connectorCount; i++) {
            const connector = createConnector();
            testConnectors.push(connector);

            const stage = extraStages[i];
            connectPromises.push(
                connector.connect(testServer.wsUrl, stage.stageId, stage.stageType)
            );
        }

        const connectResults = await Promise.all(connectPromises);

        // Then: All connections should succeed
        expect(connectResults.every(r => r === true)).toBe(true);

        // Disconnect all
        testConnectors.forEach(c => c.disconnect());
        await delay(500);

        // All should be disconnected
        expect(testConnectors.every(c => !c.isConnected)).toBe(true);
    });

    test('A-05-07: Each connector\'s events are independent', async () => {
        // Given: 2 connectors with separate event handlers
        const connector1 = createConnector();
        const connector2 = createConnector();

        const connector1Received: string[] = [];
        const connector2Received: string[] = [];

        connector1.onReceive = (packet: Packet) => {
            if (packet.msgId === 'BroadcastNotify') {
                const notify = parseBroadcastNotify(packet.payload);
                connector1Received.push(notify.data);
            }
        };

        connector2.onReceive = (packet: Packet) => {
            if (packet.msgId === 'BroadcastNotify') {
                const notify = parseBroadcastNotify(packet.payload);
                connector2Received.push(notify.data);
            }
        };

        // Connect to different stages
        await connector1.connect(testServer.wsUrl, stages[0].stageId, stages[0].stageType);
        await connector2.connect(testServer.wsUrl, stages[1].stageId, stages[1].stageType);

        await authenticate(connector1, 'connector1-user');
        await authenticate(connector2, 'connector2-user');

        // When: Each sends broadcast request
        const payload1 = serializeBroadcastRequest({ content: 'From Connector 1' });
        const broadcast1 = Packet.fromBytes('BroadcastRequest', payload1);
        connector1.send(broadcast1);

        const payload2 = serializeBroadcastRequest({ content: 'From Connector 2' });
        const broadcast2 = Packet.fromBytes('BroadcastRequest', payload2);
        connector2.send(broadcast2);

        // Wait for push messages
        const deadline = Date.now() + 5000;
        while ((connector1Received.length === 0 || connector2Received.length === 0) && Date.now() < deadline) {
            connector1.mainThreadAction();
            connector2.mainThreadAction();
            await delay(10);
        }

        // Then: Each connector receives its own messages
        expect(connector1Received.length).toBeGreaterThan(0);
        expect(connector2Received.length).toBeGreaterThan(0);
    });
});

File: connectors/cpp/tests/integration/advanced/a05_multiple_connector_test.cpp
#include "../base_integration_test.hpp"
#include <gtest/gtest.h>
#include <vector>
#include <optional>
#include <thread>
#include <chrono>

using namespace playhouse;
using namespace playhouse::test;

/// A-05: Multiple Connector Tests
/// Verifies multiple connector instances can work simultaneously
class A05_MultipleConnectorTest : public BaseIntegrationTest {};

namespace {
bool ConnectWithWait(Connector& connector, TestServerFixture& server, int timeout_ms = 5000) {
    bool connected = false;
    bool error = false;

    connector.OnConnect = [&connected]() { connected = true; };
    connector.OnError = [&error](int, std::string) { error = true; };

    auto future = connector.ConnectAsync(server.GetHost(), server.GetTcpPort());
    (void)future;

    auto deadline = std::chrono::steady_clock::now() + std::chrono::milliseconds(timeout_ms);
    while (!connected && !error && std::chrono::steady_clock::now() < deadline) {
        connector.MainThreadAction();
        std::this_thread::sleep_for(std::chrono::milliseconds(10));
    }

    return connected;
}

bool AuthenticateWithWait(Connector& connector, const std::string& user_id, int timeout_ms = 5000) {
    bool done = false;
    bool success = false;

    Bytes payload = proto::EncodeAuthenticateRequest(user_id, "valid_token");
    Packet auth_packet = Packet::FromBytes("AuthenticateRequest", std::move(payload));
    connector.Authenticate(std::move(auth_packet), [&](bool result) {
        success = result;
        done = true;
    });

    auto deadline = std::chrono::steady_clock::now() + std::chrono::milliseconds(timeout_ms);
    while (!done && std::chrono::steady_clock::now() < deadline) {
        connector.MainThreadAction();
        std::this_thread::sleep_for(std::chrono::milliseconds(10));
    }

    return done && success;
}
} // namespace

TEST_F(A05_MultipleConnectorTest, TwoConnectors_BothConnect_Independently) {
    // Given: Two connector instances
    auto connector2 = std::make_unique<Connector>();
    connector2->Init(config_);

    // When: Connect both
    auto stage1 = GetTestServer().GetOrCreateTestStage();
    auto stage2 = GetTestServer().GetOrCreateTestStage();

    bool connected1 = ConnectAndWait(5000);
    bool connected2 = ConnectWithWait(*connector2, GetTestServer(), 5000);
    ASSERT_TRUE(AuthenticateTestUser("multi_conn_user1"));
    ASSERT_TRUE(AuthenticateWithWait(*connector2, "multi_conn_user2"));

    // Then: Both should connect successfully
    EXPECT_TRUE(connected1) << "First connector should connect";
    EXPECT_TRUE(connected2) << "Second connector should connect";
    EXPECT_TRUE(connector_->IsConnected());
    EXPECT_TRUE(connector2->IsConnected());

    // Cleanup
    connector2->Disconnect();
}

TEST_F(A05_MultipleConnectorTest, MultipleConnectors_SendMessagesIndependently) {
    // Given: Multiple connectors connected
    std::vector<std::unique_ptr<Connector>> connectors;
    const int num_connectors = 3;

    for (int i = 0; i < num_connectors; i++) {
        auto conn = std::make_unique<Connector>();
        conn->Init(config_);

        auto stage = GetTestServer().GetOrCreateTestStage();
        if (ConnectWithWait(*conn, GetTestServer(), 5000)
            && AuthenticateWithWait(*conn, "multi_conn_batch_" + std::to_string(i))) {
            connectors.push_back(std::move(conn));
        }
    }

    ASSERT_GE(connectors.size(), 2) << "At least 2 connectors should connect";

    // When: Each sends a message
    std::vector<bool> completed(connectors.size(), false);
    std::vector<std::optional<Packet>> responses(connectors.size());

    for (size_t i = 0; i < connectors.size(); i++) {
        Bytes payload = proto::EncodeEchoRequest("Connector" + std::to_string(i), static_cast<int32_t>(i));
        auto packet = Packet::FromBytes("EchoRequest", std::move(payload));

        connectors[i]->Request(std::move(packet), [&, i](Packet response) {
            responses[i] = std::move(response);
            completed[i] = true;
        });
    }

    // Then: All should receive responses
    int success_count = 0;
    for (size_t i = 0; i < connectors.size(); i++) {
        auto deadline = std::chrono::steady_clock::now() + std::chrono::milliseconds(5000);
        while (!completed[i] && std::chrono::steady_clock::now() < deadline) {
            connectors[i]->MainThreadAction();
            std::this_thread::sleep_for(std::chrono::milliseconds(10));
        }
        if (completed[i] && responses[i].has_value() && responses[i]->GetErrorCode() == 0) {
            success_count++;
        }
    }

    EXPECT_GT(success_count, 0) << "At least some connectors should receive responses";

    // Cleanup
    for (auto& conn : connectors) {
        conn->Disconnect();
    }
}

TEST_F(A05_MultipleConnectorTest, Connectors_IndependentLifecycles) {
    // Given: Two connectors
    auto connector2 = std::make_unique<Connector>();
    connector2->Init(config_);

    ASSERT_TRUE(CreateStageConnectAndAuthenticate("multi_conn_lifecycle1"));

    auto stage2 = GetTestServer().GetOrCreateTestStage();
    (void)stage2;
    bool connected2 = ConnectWithWait(*connector2, GetTestServer(), 5000);
    ASSERT_TRUE(AuthenticateWithWait(*connector2, "multi_conn_lifecycle2"));

    ASSERT_TRUE(connected2);

    // When: Disconnect first connector
    connector_->Disconnect();
    std::this_thread::sleep_for(std::chrono::milliseconds(500));

    // Then: Second connector should still work
    EXPECT_FALSE(connector_->IsConnected()) << "First connector should be disconnected";
    EXPECT_TRUE(connector2->IsConnected()) << "Second connector should still be connected";

    // Cleanup
    connector2->Disconnect();
}

TEST_F(A05_MultipleConnectorTest, Connectors_SeparateCallbackHandlers) {
    // Given: Two connectors with separate callback handlers
    auto connector2 = std::make_unique<Connector>();
    connector2->Init(config_);

    std::atomic<bool> connector1_received{false};
    std::atomic<bool> connector2_received{false};

    connector_->OnReceive = [&](Packet packet) {
        connector1_received = true;
    };

    connector2->OnReceive = [&](Packet packet) {
        connector2_received = true;
    };

    // Connect both
    ASSERT_TRUE(CreateStageConnectAndAuthenticate("multi_conn_cb1"));

    auto stage2 = GetTestServer().GetOrCreateTestStage();
    (void)stage2;
    ASSERT_TRUE(ConnectWithWait(*connector2, GetTestServer(), 5000));
    ASSERT_TRUE(AuthenticateWithWait(*connector2, "multi_conn_cb2"));

    // When: Trigger messages for both
    Bytes payload1 = proto::EncodeBroadcastRequest("connector1");
    connector_->Send(Packet::FromBytes("BroadcastRequest", std::move(payload1)));

    Bytes payload2 = proto::EncodeBroadcastRequest("connector2");
    connector2->Send(Packet::FromBytes("BroadcastRequest", std::move(payload2)));

    // Process callbacks for both
    for (int i = 0; i < 50; i++) {
        connector_->MainThreadAction();
        connector2->MainThreadAction();
        std::this_thread::sleep_for(std::chrono::milliseconds(10));

        if (connector1_received && connector2_received) break;
    }

    // Then: Each should receive their own callback
    EXPECT_TRUE(connector1_received || connector2_received) << "At least one connector should receive callback";

    // Cleanup
    connector2->Disconnect();
}

TEST_F(A05_MultipleConnectorTest, Connectors_SequentialCreationAndDestruction) {
    // When: Create and destroy connectors sequentially
    for (int i = 0; i < 3; i++) {
        auto conn = std::make_unique<Connector>();
        conn->Init(config_);

        auto stage = GetTestServer().GetOrCreateTestStage();
        (void)stage;
        bool connected = ConnectWithWait(*conn, GetTestServer(), 5000);
        if (connected && AuthenticateWithWait(*conn, "multi_conn_seq_" + std::to_string(i))) {
            conn->Disconnect();
        }
        // conn destroyed here
    }

    // Then: Should handle sequential lifecycle without issues
    SUCCEED() << "Sequential connector creation/destruction handled correctly";
}
